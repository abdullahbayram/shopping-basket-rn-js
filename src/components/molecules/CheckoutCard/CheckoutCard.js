import * as React from 'react';
import PropTypes from 'prop-types';
import { View, Image } from 'react-native';
import { MaterialIcons } from '@expo/vector-icons';
import Button from '../../atoms/Button/Button';
import Text from '../../atoms/Text/Text';
import showToast from '../../../utils/showToast';
import messages from '../../../constants/strings';
import styles from './CheckoutCard.style';

const CheckoutCard = ({ product, onQuantityChange, onRemoveButtonPress }) => {
  const [localQuantity, setLocalQuantity] = React.useState(product.quantity);

  const isQuantityEqualsToOne = localQuantity === 1;
  const leftButtonIcon = isQuantityEqualsToOne ? 'delete' : 'remove';

  const handleDecrease = () => {
    if (localQuantity > 1) {
      setLocalQuantity(localQuantity - 1);
      onQuantityChange(product, localQuantity - 1);
    } else {
      onRemoveButtonPress();
    }
  };

  const handleIncrease = () => {
    if (localQuantity < 5) {
      setLocalQuantity(localQuantity + 1);
      onQuantityChange(product, localQuantity + 1);
    } else {
      showToast(messages.invalidQuantity);
    }
  };

  const formattedPrice = (product.price * localQuantity).toFixed(2);

  return (
    <View style={styles.container}>
      <View style={styles.topSection}>
        <Image style={styles.image} source={{ uri: product.image }} />
        <View style={styles.infoContainer}>
          <Text style={styles.title} numberOfLines={2} ellipsizeMode="tail">
            {product.title}
          </Text>
          <Text style={styles.subtitle} numberOfLines={2} ellipsizeMode="tail">
            {product.description}
          </Text>
          <Text style={styles.price}>${formattedPrice}</Text>
        </View>
      </View>

      <View style={styles.bottomSection}>
        <View style={styles.quantityContainer}>
          <Button onPress={handleDecrease} style={styles.quantityMinButton}>
            <MaterialIcons name={leftButtonIcon} size={18} color="#FFA500" />
          </Button>
          <View style={styles.quantityTextContainer}>
            <Text style={styles.quantityText}>{localQuantity}</Text>
          </View>
          <Button onPress={handleIncrease} style={styles.quantityPlusButton}>
            <MaterialIcons name="add" size={18} color="#FFA500" />
          </Button>
        </View>
        <Button onPress={onRemoveButtonPress} style={styles.transparentButton}>
          Remove Item
        </Button>
      </View>
    </View>
  );
};

CheckoutCard.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    quantity: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
  }).isRequired,
  onQuantityChange: PropTypes.func.isRequired,
  onRemoveButtonPress: PropTypes.func.isRequired,
};

export default CheckoutCard;
