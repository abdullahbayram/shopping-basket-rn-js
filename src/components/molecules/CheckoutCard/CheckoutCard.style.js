import { StyleSheet } from 'react-native';
import { typography } from '@constants/theme';

export default (colors, spacing) =>
  StyleSheet.create({
    container: {
      backgroundColor: colors.cardBackground,
      marginVertical: spacing.sm,
      borderRadius: spacing.sm,
      padding: spacing.sm,
      shadowColor: colors.shadowColor,
      shadowOffset: { width: 0, height: 2 },
      shadowOpacity: 0.2,
      shadowRadius: 4,
      elevation: 4,
    },
    topSection: {
      flexDirection: 'row',
      marginBottom: spacing.md,
    },
    image: {
      width: 100,
      height: 100,
      borderRadius: 8,
      resizeMode: 'cover',
    },
    infoContainer: {
      flex: 1,
      paddingHorizontal: spacing.sm,
    },
    title: {
      fontSize: typography.default,
      fontWeight: typography.bold,
      marginBottom: spacing.sm,
      color: colors.textTertiary,
    },
    subtitle: {
      fontSize: typography.md,
      color: colors.textSecondary,
      marginBottom: spacing.special,
    },
    price: {
      fontSize: typography.default,
      fontWeight: typography.bold,
      color: colors.textTertiary,
    },
    bottomSection: {
      flexDirection: 'row',
      flex: 1,
    },
    quantityContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      flex: 1,
    },
    quantityButton: {
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: colors.buttonBackground,
      borderWidth: 1,
      borderColor: colors.buttonBorder,
      borderRadius: 20,
      maxWidth: 50,
    },
    quantityTextContainer: {
      width: 30,
    },
    quantityText: {
      fontSize: typography.default,
      fontWeight: typography.bold,
      textAlign: 'center',
      color: colors.textTertiary,
    },
    transparentButton: {
      backgroundColor: 'transparent',
      borderWidth: 1,
      borderColor: colors.transparentButtonBorder,
      borderRadius: 20,
    },
  });
