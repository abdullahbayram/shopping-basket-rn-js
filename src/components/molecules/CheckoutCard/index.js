import React from 'react';
import PropTypes from 'prop-types';
import { View, Image } from 'react-native';
import { MaterialIcons } from '@expo/vector-icons';
import { useTheme } from 'react-native-paper';
import messages from '@constants/toastMessages';
import { spacing } from '@constants/theme';
import { showToast } from '@utils';
import { Button, Text } from '../../atoms';
import createStyles from './CheckoutCard.style';

const CheckoutCard = ({ product, maxQuantity = 5, onQuantityChange, onRemoveButtonPress }) => {
  const { colors } = useTheme();
  const styles = createStyles(colors, spacing);

  const isQuantityEqualsToOne = product.quantity === 1;
  const formattedPrice = (product.price * product.quantity).toFixed(2);

  const handleDecrease = () => {
    if (product.quantity > 1) {
      onQuantityChange(product, product.quantity - 1);
    } else {
      onRemoveButtonPress();
    }
  };

  const handleIncrease = () => {
    if (product.quantity < maxQuantity) {
      onQuantityChange(product, product.quantity + 1);
    } else {
      showToast(messages.basket.limitReached);
    }
  };

  return (
    <View testID="checkout-card" style={styles.container}>
      <View style={styles.topSection}>
        <Image style={styles.image} source={{ uri: product.image }} />
        <View style={styles.infoContainer}>
          <Text style={styles.title} numberOfLines={2} ellipsizeMode="tail">
            {product.title}
          </Text>
          <Text style={styles.subtitle} numberOfLines={2} ellipsizeMode="tail">
            {product.description}
          </Text>
          <Text style={styles.price}>${formattedPrice}</Text>
        </View>
      </View>

      <View style={styles.bottomSection}>
        <View style={styles.quantityContainer}>
          <Button
            testID={isQuantityEqualsToOne ? 'delete-button' : 'decrease-button'}
            onPress={handleDecrease}
            style={styles.quantityButton}
          >
            <MaterialIcons name={isQuantityEqualsToOne ? 'delete' : 'remove'} size={18} color={colors.buttonBorder} />
          </Button>
          <View style={styles.quantityTextContainer}>
            <Text testID="product-quantity" style={styles.quantityText}>
              {product.quantity}
            </Text>
          </View>
          <Button testID="increase-button" onPress={handleIncrease} style={styles.quantityButton}>
            <MaterialIcons name="add" size={18} color={colors.buttonBorder} />
          </Button>
        </View>
        <Button onPress={onRemoveButtonPress} style={styles.transparentButton}>
          Remove Item
        </Button>
      </View>
    </View>
  );
};

CheckoutCard.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    quantity: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
  }).isRequired,
  maxQuantity: PropTypes.number,
  onQuantityChange: PropTypes.func.isRequired,
  onRemoveButtonPress: PropTypes.func.isRequired,
};

export default CheckoutCard;
