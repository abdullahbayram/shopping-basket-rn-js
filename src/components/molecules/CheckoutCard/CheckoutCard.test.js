import React from 'react';
import { fireEvent, screen } from '@testing-library/react-native';
import { showToast } from '@utils';
import { renderInThemeProvider } from '@testUtils/renderInThemeProvider';
import CheckoutCard from '.';

jest.mock('@utils', () => ({
  showToast: jest.fn(),
}));

jest.mock('./CheckoutCard.style', () => jest.fn(() => ({})));

describe('CheckoutCard', () => {
  const mockProduct = {
    id: 1,
    quantity: 2,
    price: 10.0,
    title: 'Test Product',
    description: 'This is a test product.',
    image: 'https://example.com/image.jpg',
  };

  const mockOnQuantityChange = jest.fn();
  const mockOnRemoveButtonPress = jest.fn();

  const renderComponent = (props) =>
    renderInThemeProvider(
      <CheckoutCard
        product={mockProduct}
        maxQuantity={5}
        onQuantityChange={mockOnQuantityChange}
        onRemoveButtonPress={mockOnRemoveButtonPress}
        {...props}
      />,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly and matches the snapshot', () => {
    const { toJSON } = renderComponent();
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onQuantityChange with decreased quantity when decrease button is pressed', () => {
    renderComponent();

    const decreaseButton = screen.getByTestId('decrease-button');
    fireEvent.press(decreaseButton);

    expect(mockOnQuantityChange).toHaveBeenCalledWith(mockProduct, 1);
  });

  it('calls onRemoveButtonPress when quantity is 1 and decrease button is pressed', () => {
    renderComponent({
      product: { ...mockProduct, quantity: 1 },
    });

    const removeButton = screen.getByText('Remove Item');
    fireEvent.press(removeButton);

    expect(mockOnRemoveButtonPress).toHaveBeenCalled();
  });

  it('calls onQuantityChange with increased quantity when increase button is pressed', () => {
    renderComponent();

    const increaseButton = screen.getByTestId('increase-button');
    fireEvent.press(increaseButton);

    expect(mockOnQuantityChange).toHaveBeenCalledWith(mockProduct, 3);
  });

  it('shows a toast when trying to increase quantity above maxQuantity', () => {
    renderComponent({
      product: { ...mockProduct, quantity: 5 },
    });

    const increaseButton = screen.getByTestId('increase-button');
    fireEvent.press(increaseButton);

    expect(showToast).toHaveBeenCalledWith({ msg: 'Quantity must be between 1 and 5.', title: 'Invalid Quantity' });
  });

  it('calls onRemoveButtonPress when Remove Item button is pressed', () => {
    renderComponent();

    const removeItemButton = screen.getByText('Remove Item');
    fireEvent.press(removeItemButton);

    expect(mockOnRemoveButtonPress).toHaveBeenCalled();
  });

  it('renders the correct formatted price', () => {
    renderComponent();

    const priceText = screen.getByText('$20.00');
    expect(priceText).toBeTruthy();
  });
});
