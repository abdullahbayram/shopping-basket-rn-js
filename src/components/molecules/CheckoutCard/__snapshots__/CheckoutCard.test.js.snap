// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutCard renders correctly and matches the snapshot 1`] = `
<View>
  <View>
    <Image
      source={
        {
          "uri": "https://example.com/image.jpg",
        }
      }
    />
    <View>
      <Text
        ellipsizeMode="tail"
        numberOfLines={2}
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              [
                {
                  "color": "#34300F",
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Test Product
      </Text>
      <Text
        ellipsizeMode="tail"
        numberOfLines={2}
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              [
                {
                  "color": "#34300F",
                },
                undefined,
              ],
            ],
          ]
        }
      >
        This is a test product.
      </Text>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              [
                {
                  "color": "#34300F",
                },
                undefined,
              ],
            ],
          ]
        }
      >
        $
        20.00
      </Text>
    </View>
  </View>
  <View>
    <View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#FED034",
            "borderRadius": 10,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="decrease-button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#FED034",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 10,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="decrease-button-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 10,
                },
              ]
            }
            testID="decrease-button"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "#34300F",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="decrease-button-text"
              >
                <Text />
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                [
                  {
                    "color": "#34300F",
                  },
                  undefined,
                ],
              ],
            ]
          }
        >
          2
        </Text>
      </View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#FED034",
            "borderRadius": 10,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="increase-button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#FED034",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 10,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="increase-button-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 10,
                },
              ]
            }
            testID="increase-button"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "#34300F",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="increase-button-text"
              >
                <Text />
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FED034",
          "borderRadius": 10,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#FED034",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 10,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flex": undefined,
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 10,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      undefined,
                      false,
                      {
                        "color": "#34300F",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Remove Item
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
